File: go/solutions/0343-integer-break.go
// https://leetcode.com/problems/integer-break/
package solutions

import "github.com/SamuelMuloki/LeetCode/go/utils"

func IntegerBreak(n int) int {
	if n <= 3 {
		return n - 1
	}

	dp := make([]int, n+1)
	var dfs func(num int) int
	dfs = func(num int) int {
		if num <= 3 {
			return num
		}

		if dp[num] != 0 {
			return dp[num]
		}

		res := num
		for i := 2; i < num; i++ {
			res = utils.Max(res, i*dfs(num-i))
		}

		dp[num] = res
		return dp[num]
	}

	return dfs(n)
}

File: js/0343-integer-break.js
/**
 * @param {number} n
 * @return {number}
 */
var integerBreak = function (n) {
  if (n <= 3) {
    return n - 1;
  }

  var dp = new Array(n + 1);
  var dfs = function (num) {
    if (num <= 3) {
      return num;
    }

    if (dp[num]) {
      return dp[num];
    }

    var res = num;
    for (var i = 2; i < num; i++) {
      res = Math.max(res, i * dfs(num - i));
    }

    return (dp[num] = res);
  };

  return dfs(n);
};

module.exports = { integerBreak };
